### PR TITLE
Decorate emitters with PublisherDecorators

### LINKED ISSUE #2332: Missing metrics for channels associated with Emitters
We have created an emitter that is sending messages to a kafka deployment via our own connector and we are pulling messages from Kafka on a separate channel. When we look at the metrics that are being produced. We have a single entry for mp_messaging that is for the incoming channel. But we are not seeing an entry at all the emitter channel.

When reviewing the [specs](https://download.eclipse.org/microprofile/microprofile-reactive-messaging-3.0/microprofile-reactive-messaging-spec-3.0.html#_metrics) for metrics. This part: `metrics are produced for each channel declared by the application`. There is no statement about how a channel associated with an emitter.

Our scenario can be seen as the following simple flows:
Emitter -> channelA -> Kafka connector -> Kafka
Kafka -> Kafka Connector -> channelB -> `@Incoming`

We are seeing metrics only for channelB.

When we used an `@Outgoing` and `@Incoming` we got metrics to both. so would indicate that it is the fact we had the xhannel  associated with an Emitter that meant we were not seeing any metrics.

### PR DESCRIPTION
**Pull request by ozangunalp**
Resolves #2332 
May supersede #2336

### CODE REVIEW

File: smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/providers/wiring/Wiring.java [hunk 1]
```diff
@@ -504,7 +515,13 @@ public void materialize(ChannelRegistry registry) {
         private <T extends MessagePublisherProvider<?>> void registerEmitter(ChannelRegistry registry, int def) {
             EmitterFactory<?> emitterFactory = getEmitterFactory(configuration.emitterType());
             T emitter = (T) emitterFactory.createEmitter(configuration, def);
-            Publisher<? extends Message<?>> publisher = emitter.getPublisher();
+            Multi<? extends Message<?>> publisher = Multi.createFrom().publisher(emitter.getPublisher());
+            for (PublisherDecorator decorator : getSortedInstances(decorators)) {
+                // TODO do not apply ContextDecorator here, it messes with custom emitter transactions
```
Comment (cescoffier) [inline smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/providers/wiring/Wiring.java:520]: Any reason why?

Comment (cescoffier) [reply smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/providers/wiring/Wiring.java:520]: If we do not apply it, how to we associate the context to the message?

File: smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/providers/wiring/Wiring.java [hunk 2]
```diff
@@ -504,7 +515,13 @@ public void materialize(ChannelRegistry registry) {
...
+                if (!(decorator instanceof ContextDecorator)) {
+                    publisher = decorator.decorate(publisher, configuration.name(), false);
```
Comment (cescoffier) [inline smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/providers/wiring/Wiring.java:522]: So, you consider the emitter as a "not connector", which is exact. However, I was wondering if we need to be specific in this case and introduce a second flag (I'm not a big fan of signatures containing two exclusive flags, having an enum would be better).

However, do we need to know when it's an emitter?

**Review by cescoffier — CHANGES_REQUESTED** [review]

**Comment by ozangunalp** [comment]
@cescoffier pushed a fix for Kafka transactions. It appears that applying ContextDecorator there changes the tx flow order, flushing and commiting the Tx before records are sent. Now it captures sent record Uni's and makes sure that all of them are sent before flusing and committing.

**Comment by ozangunalp** [comment]
@cescoffier shall we get this in ?

**Comment by cescoffier** [comment]
Yes!

**Pull request merged**